ambari-utility/src/main/java/org/apache/ambari/swagger/AmbariSwaggerReader.java

```diff
@@ -84,32 +85,54 @@ public Swagger read(Set<Class<?>> classes) {
                   returnType.getName());
             }
             else {
-              Path apiPath = AnnotationUtils.findAnnotation(cls, Path.class);
-              String apiPathValue;
-              if (null == apiPath) {
-                logger.warn("Parent api {} also seems to be a nested API. The current version does not support " +
-                    "multi-level nesting.");
-                apiPathValue = "";
-              }
-              else {
-                apiPathValue = apiPath.value();
-              }
-              NestedApiRecord nar = new NestedApiRecord(returnType, cls, apiPathValue, method, methodPath.value());
+              Boolean skipAdd = false;
+              Class<?> preferredParentClass = cls;
+
+              // API is a nested API of multiple top level APIs
               if (nestedAPIs.containsKey(returnType)) {
-                logger.warn("{} is a nested API of multiple top level API's. Ignoring top level API {}", returnType, cls);
+                SwaggerPreferredParent preferredParentAnnotation = AnnotationUtils.findAnnotation(returnType,
+                        SwaggerPreferredParent.class);
+                if (null != preferredParentAnnotation) {
+                  preferredParentClass = preferredParentAnnotation.preferredParent();
+                  if (nestedAPIs.get(returnType).parentApi.getName().equals(preferredParentClass.getName())) {
+                    skipAdd = true;
+                  } else {
+                    logger.info("Setting top level API of {} to {} based on @SwaggerPreferredParent " +
+                            "annotation", returnType, preferredParentClass.getSimpleName());
+                    try {
+                      method = preferredParentClass.getMethod(method.getName(), method.getParameterTypes());
+                    } catch (NoSuchMethodException exc) {
+                      skipAdd = true;
+                      logger.error("{} class defined as parent API is invalid due to method mismatch! Ignoring " +
+                              "API {}", preferredParentClass, returnType);
+                    }
+                  }
+                } else {
+                  logger.warn("{} is a nested API of multiple top level API's. Ignoring top level API {}",
+                          returnType, cls);
+                  skipAdd = true;
+                }
               }
-              else {
-                logger.info("Registering nested API: {}", returnType);
-                nestedAPIs.put(returnType, nar);
+
+              if (skipAdd) {
+                continue;
+              } else {
+                nestedAPIs.remove(returnType);
+                cls = preferredParentClass;
               }
+
+              logger.info("Registering nested API: {}", returnType);
+              NestedApiRecord nar = new NestedApiRecord(returnType, cls, validateParentApiPath(cls), method,
```

Comment:
For better readability I would skip the assignment in line 129 and would use preferredParentClass instead of cls in line 133.